File: packages/cli/src/state/progressCommit.ts
import { ancestorState } from "../core/git.js";
import { loadYamlMapping } from "../core/yaml.js";

/**
 * Shared progress-commit and git-ancestry logic. Faithful TS port of
 * `scripts/progress_commit.py`.
 */

export const COMMIT_HASH_RE = /^[0-9a-fA-F]{7,40}$/;

/**
 * Leading verifiable git hash in a cycle `commit` value, else `null`. `pending`,
 * `N/A …` and non-hash free text are exempt; non-string values return `null`.
 */
export function commitToken(value: unknown): string | null {
  if (typeof value !== "string") {
    return null;
  }
  const text = value.trim();
  if (!text) {
    return null;
  }
  const token = text.split(/\s+/)[0];
  const low = token.toLowerCase();
  if (low === "pending" || low.startsWith("n/a")) {
    return null;
  }
  return COMMIT_HASH_RE.test(token) ? token : null;
}

type ChangeMap = Map<number, string> | Record<number, string>;

function changeHas(changes: ChangeMap, key: number): boolean {
  return changes instanceof Map ? changes.has(key) : Object.prototype.hasOwnProperty.call(changes, key);
}

function changeGet(changes: ChangeMap, key: number): string {
  return changes instanceof Map ? (changes.get(key) as string) : (changes as Record<number, string>)[key];
}

function changeSize(changes: ChangeMap): number {
  return changes instanceof Map ? changes.size : Object.keys(changes).length;
}

function splitKeepEnds(text: string): string[] {
  const matches = text.match(/[^\n]*\n|[^\n]+/g);
  return matches ?? [];
}

/**
 * Replace the `commit:` field of specific cycles, preserving the rest. `changes`
 * maps cycle number to the new `commit` value. Multi-line commit scalars are
 * replaced wholesale, dropping their continuation lines.
 */
export function rewriteCycleCommits(text: string, changes: ChangeMap): string {
  if (changeSize(changes) === 0) {
    return text;
  }
  const numberRe = /^-\s+number:\s*(\d+)\s*$/;
  const listItemRe = /^-\s/;
  const commitRe = /^(\s+)commit:\s*.*$/;
  const topKeyRe = /^[A-Za-z_]/;
  const lines = splitKeepEnds(text);
  const out: string[] = [];
  let inCycles = false;
  let current: number | null = null;
  let i = 0;
  const total = lines.length;
  while (i < total) {
    const line = lines[i];
    const body = line.endsWith("\n") ? line.slice(0, -1) : line;
    if (topKeyRe.test(body)) {
      inCycles = body.startsWith("cycles:");
      current = null;
      out.push(line);
      i += 1;
      continue;
    }
    if (inCycles) {
      const numberMatch = numberRe.exec(body);
      if (numberMatch) {
        current = parseInt(numberMatch[1], 10);
        out.push(line);
        i += 1;
        continue;
      }
      if (listItemRe.test(body)) {
        current = null;
        out.push(line);
        i += 1;
        continue;
      }
      const commitMatch = commitRe.exec(body);
      if (commitMatch && current !== null && changeHas(changes, current)) {
        const indent = commitMatch[1];
        const keyIndent = indent.length;
        const newline = line.endsWith("\n") ? "\n" : "";
        out.push(`${indent}commit: ${changeGet(changes, current)}${newline}`);
        i += 1;
        while (i < total) {
          const cont = lines[i].endsWith("\n") ? lines[i].slice(0, -1) : lines[i];
          if (!cont.trim()) {
            break;
          }
          const leading = (cont.match(/^ */)?.[0].length) ?? 0;
          if (leading > keyIndent && !listItemRe.test(cont)) {
            i += 1;
            continue;
          }
          break;
        }
        current = null;
        continue;
      }
    }
    out.push(line);
    i += 1;
  }
  return out.join("");
}

export interface BackfillOperation {
  cycle: number;
  commit: unknown;
  state: string;
  action: string;
}

export interface BackfillResult {
  status: string;
  exitCode: number;
  operations: BackfillOperation[];
  /** Descending-by-cycle ordered list of cycle->value rewrites. */
  changes: Array<[number, string]>;
  message: string | null;
}

export interface BackfillOptions {
  mode?: string;
  targetCommit?: string | null;
  targetCycle?: number | null;
  cwd?: string;
}

/**
 * Classify progress cycle commits and decide the backfill outcome. Pure with
 * respect to the artifact: parses `text` and inspects git ancestry but never
 * reads or writes the progress file.
 */
export function computeBackfill(text: string, options: BackfillOptions = {}): BackfillResult {
  const mode = options.mode ?? "check";
  const targetCommit = options.targetCommit ?? null;
  let targetCycle = options.targetCycle ?? null;
  const cwd = options.cwd ?? ".";

  let data: Record<string, unknown>;
  try {
    data = loadYamlMapping(text);
  } catch (exc) {
    return {
      status: "error",
      exitCode: 2,
      operations: [],
      changes: [],
      message: `cannot parse progress artifact: ${(exc as Error).message}`,
    };
  }
  const cyclesRaw = data && typeof data === "object" ? (data as Record<string, unknown>).cycles : null;
  if (!Array.isArray(cyclesRaw)) {
    return {
      status: "noop",
      exitCode: 0,
      operations: [],
      changes: [],
      message: "progress artifact has no cycles list",
    };
  }

  if (ancestorState("HEAD", cwd) === "unavailable") {
    return {
      status: "noop",
      exitCode: 0,
      operations: [],
      changes: [],
      message: "git HEAD unavailable; cannot verify commit ancestry",
    };
  }

  const fullCycles: Array<[number, unknown]> = [];
  for (const c of cyclesRaw) {
    if (c && typeof c === "object" && !Array.isArray(c)) {
      const num = (c as Record<string, unknown>).number;
      if (typeof num === "number" && Number.isInteger(num)) {
        fullCycles.push([num, (c as Record<string, unknown>).commit]);
      }
    }
  }

  let targetToken: string | null = null;
  if (targetCommit) {
    const trimmed = targetCommit.trim();
    targetToken = trimmed ? trimmed.split(/\s+/)[0] : "";
    const state = targetToken ? ancestorState(targetToken, cwd) : "unknown";
    if (state !== "ancestor") {
      return {
        status: "error",
        exitCode: 2,
        operations: [],
        changes: [],
        message:
          `refusing to backfill commit '${targetCommit}': it is ${state} relative to HEAD. ` +
          "Backfill only a commit already in HEAD's history; commit the product change first, " +
          "then backfill from a later commit (never amend to backfill).",
      };
    }
    const known = new Set(fullCycles.map(([n]) => n));
    if (targetCycle !== null && !known.has(targetCycle)) {
      return {
        status: "error",
        exitCode: 2,
        operations: [],
        changes: [],
        message: `cycle ${targetCycle} not found in progress cycles`,
      };
    }
    if (targetCycle === null) {
      targetCycle =
        fullCycles.find(([, cv]) => {
          const tok = commitToken(cv);
          return tok === null || ancestorState(tok ?? "", cwd) === "stale";
        })?.[0] ?? null;
    }
    if (targetCycle === null) {
      return {
        status: "noop",
        exitCode: 0,
        operations: [],
        changes: [],
        message: "no pending or stale cycle to backfill; pass --cycle N to target one",
      };
    }
  }

  const changes = new Map<number, string>();
  const operations: BackfillOperation[] = [];
  for (const [number, commitValue] of fullCycles) {
    if (number === targetCycle && targetToken) {
      let action = "set-commit";
      let state = "ancestor";
      if (commitToken(commitValue) === targetToken) {
        action = "none";
        state = "ancestor";
      } else {
        changes.set(number, targetToken);
      }
      operations.push({ cycle: number, commit: targetToken, state, action });
      continue;
    }
    const token = commitToken(commitValue);
    if (token === null) {
      operations.push({ cycle: number, commit: commitValue, state: "exempt", action: "none" });
      continue;
    }
    const state = ancestorState(token, cwd);
    if (state === "stale") {
      changes.set(number, "pending");
      operations.push({ cycle: number, commit: token, state: "stale", action: "reset-to-pending" });
    } else {
      operations.push({ cycle: number, commit: token, state, action: "none" });
    }
  }

  const sortedChanges: Array<[number, string]> = [...changes.entries()].sort((a, b) => b[0] - a[0]);
  if (changes.size === 0) {
    return { status: "clean", exitCode: 0, operations, changes: sortedChanges, message: null };
  }
  if (mode === "fix") {
    return { status: "fixed", exitCode: 0, operations, changes: sortedChanges, message: null };
  }
  return { status: "action-needed", exitCode: 1, operations, changes: sortedChanges, message: null };
}

/**
 * Flag progress cycle `commit` hashes that are not ancestors of HEAD. Faithful
 * port of progress_commit.validate_progress_commits.
 */
export function validateProgressCommits(content: string, cwd: string = "."): string[] {
  let data: Record<string, unknown>;
  try {
    data = loadYamlMapping(content);
  } catch {
    return [];
  }
  if (!data || typeof data !== "object" || Array.isArray(data)) {
    return [];
  }
  const cycles = (data as Record<string, unknown>).cycles;
  if (!Array.isArray(cycles)) {
    return [];
  }
  if (ancestorState("HEAD", cwd) === "unavailable") {
    return [];
  }
  const violations: string[] = [];
  for (const entry of cycles) {
    if (!entry || typeof entry !== "object" || Array.isArray(entry)) {
      continue;
    }
    const raw = (entry as Record<string, unknown>).commit;
    if (typeof raw !== "string") {
      continue;
    }
    const token = commitToken(raw);
    if (token === null) {
      continue;
    }
    if (ancestorState(token, cwd) === "stale") {
      const number = (entry as Record<string, unknown>).number ?? "?";
      violations.push(
        `progress: cycle ${number} commit '${token}' is not an ancestor of HEAD ` +
          "(stale or self-referential); set it to `pending` or run " +
          "`agentera check backfill --mode fix`, then forward-fill the product commit " +
          "(never amend to backfill)",
      );
    }
  }
  return violations;
}

File: scripts/progress_commit.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = ["pyyaml"]
# ///
"""Shared progress-commit and git-ancestry logic for Agentera.

One implementation of the commit-hash regex, git subprocess wrapper, cycle
``commit`` token parser, HEAD-ancestor classifier, cycle-commit rewrite, and the
pure backfill computation. The ``agentera check backfill`` CLI command consumes
this module so the logic lives in exactly one place.

The module imports only the Python standard library plus the sibling
``yaml_mapping`` wrapper, so it is safe to import from the hook layer (no
third-party dependency beyond the sanctioned pyyaml wrapper).
"""

from __future__ import annotations

import re
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from yaml_mapping import load_yaml_mapping

COMMIT_HASH_RE = re.compile(r"^[0-9a-fA-F]{7,40}$")


def git_run(args: list[str], cwd: Path | str) -> subprocess.CompletedProcess[str] | None:
    """Run ``git`` in ``cwd``; return ``None`` if git cannot be invoked."""
    try:
        return subprocess.run(
            ["git", *args],
            cwd=str(cwd),
            capture_output=True,
            text=True,
            check=False,
        )
    except (OSError, subprocess.SubprocessError):
        return None


def commit_token(value: Any) -> str | None:
    """Leading verifiable git hash in a cycle ``commit`` value, else ``None``.

    ``pending``, ``N/A …`` and non-hash free text are exempt (return ``None``),
    as are non-string values.
    """
    if not isinstance(value, str):
        return None
    text = value.strip()
    if not text:
        return None
    token = text.split()[0]
    low = token.lower()
    if low == "pending" or low.startswith("n/a"):
        return None
    return token if COMMIT_HASH_RE.match(token) else None


def ancestor_state(token: str, cwd: Path | str) -> str:
    """Classify ``token`` relative to HEAD.

    One of ``ancestor`` (already in history), ``stale`` (resolves but not an
    ancestor of HEAD), ``unknown`` (not a commit object here), or
    ``unavailable`` (git/HEAD missing).
    """
    head = git_run(["rev-parse", "--verify", "--quiet", "HEAD"], cwd)
    if head is None or head.returncode != 0:
        return "unavailable"
    exists = git_run(["rev-parse", "--verify", "--quiet", f"{token}^{{commit}}"], cwd)
    if exists is None:
        return "unavailable"
    if exists.returncode != 0:
        return "unknown"
    ancestor = git_run(["merge-base", "--is-ancestor", token, "HEAD"], cwd)
    if ancestor is None:
        return "unavailable"
    if ancestor.returncode == 0:
        return "ancestor"
    if ancestor.returncode == 1:
        return "stale"
    return "unavailable"


def rewrite_cycle_commits(text: str, changes: dict[int, str]) -> str:
    """Replace the ``commit:`` field of specific cycles, preserving the rest.

    ``changes`` maps cycle number to the new ``commit`` value. Only entries in
    the top-level ``cycles:`` list are touched; comments, ordering, quoting, and
    every other field are left exactly as authored. Multi-line commit scalars
    (a hash wrapped with a subject across several deeper-indented lines) are
    replaced wholesale, dropping their continuation lines.
    """
    if not changes:
        return text
    number_re = re.compile(r"^-\s+number:\s*(\d+)\s*$")
    list_item_re = re.compile(r"^-\s")
    commit_re = re.compile(r"^(\s+)commit:\s*.*$")
    top_key_re = re.compile(r"^[A-Za-z_]")
    lines = text.splitlines(keepends=True)
    out: list[str] = []
    in_cycles = False
    current: int | None = None
    i = 0
    total = len(lines)
    while i < total:
        line = lines[i]
        body = line.rstrip("\n")
        if top_key_re.match(body):
            in_cycles = body.startswith("cycles:")
            current = None
            out.append(line)
            i += 1
            continue
        if in_cycles:
            number_match = number_re.match(body)
            if number_match:
                current = int(number_match.group(1))
                out.append(line)
                i += 1
                continue
            if list_item_re.match(body):
                current = None
                out.append(line)
                i += 1
                continue
            commit_match = commit_re.match(body)
            if commit_match and current in changes:
                indent = commit_match.group(1)
                key_indent = len(indent)
                newline = "\n" if line.endswith("\n") else ""
                out.append(f"{indent}commit: {changes[current]}{newline}")
                i += 1
                while i < total:
                    cont = lines[i].rstrip("\n")
                    if not cont.strip():
                        break
                    leading = len(cont) - len(cont.lstrip(" "))
                    if leading > key_indent and not list_item_re.match(cont):
                        i += 1
                        continue
                    break
                current = None
                continue
        out.append(line)
        i += 1
    return "".join(out)


@dataclass(frozen=True)
class BackfillResult:
    """Outcome of a pure backfill computation.

    ``status`` and ``exit_code`` mirror the observable CLI contract
    (``clean``/``action-needed``/``fixed``/``error``/``noop`` and exit codes
    0/1/2). ``operations`` and ``changes`` describe the per-cycle classification
    and the cycle→value rewrites the caller should apply when ``status`` is
    ``fixed``. ``message`` is set only for noop/error explanations.
    """

    status: str
    exit_code: int
    operations: list[dict[str, Any]]
    changes: dict[int, str]
    message: str | None = None


def compute_backfill(
    text: str,
    *,
    mode: str = "check",
    target_commit: str | None = None,
    target_cycle: int | None = None,
    cwd: Path | str = ".",
) -> BackfillResult:
    """Classify progress cycle commits and decide the backfill outcome.

    Pure with respect to the artifact: it parses ``text`` and inspects git
    ancestry but never reads or writes the progress file. The caller performs
    file I/O and applies :func:`rewrite_cycle_commits` when ``status`` is
    ``fixed``.
    """
    try:
        data = load_yaml_mapping(text)
    except Exception as exc:  # noqa: BLE001 - mirror CLI: any load failure is an error
        return BackfillResult("error", 2, [], {}, f"cannot parse progress artifact: {exc}")
    cycles = data.get("cycles") if isinstance(data, dict) else None
    if not isinstance(cycles, list):
        return BackfillResult("noop", 0, [], {}, "progress artifact has no cycles list")

    if ancestor_state("HEAD", cwd) == "unavailable":
        return BackfillResult(
            "noop", 0, [], {}, "git HEAD unavailable; cannot verify commit ancestry"
        )

    full_cycles = [
        (c["number"], c.get("commit"))
        for c in cycles
        if isinstance(c, dict) and isinstance(c.get("number"), int)
    ]

    target_token: str | None = None
    if target_commit:
        target_token = target_commit.strip().split()[0] if target_commit.strip() else ""
        state = ancestor_state(target_token, cwd) if target_token else "unknown"
        if state != "ancestor":
            return BackfillResult(
                "error",
                2,
                [],
                {},
                f"refusing to backfill commit '{target_commit}': it is {state} relative to HEAD. "
                "Backfill only a commit already in HEAD's history; commit the product change first, "
                "then backfill from a later commit (never amend to backfill).",
            )
        known = {n for n, _ in full_cycles}
        if target_cycle is not None and target_cycle not in known:
            return BackfillResult(
                "error", 2, [], {}, f"cycle {target_cycle} not found in progress cycles"
            )
        if target_cycle is None:
            target_cycle = next(
                (
                    n
                    for n, cv in full_cycles
                    if commit_token(cv) is None
                    or ancestor_state(commit_token(cv) or "", cwd) == "stale"
                ),
                None,
            )
        if target_cycle is None:
            return BackfillResult(
                "noop",
                0,
                [],
                {},
                "no pending or stale cycle to backfill; pass --cycle N to target one",
            )

    changes: dict[int, str] = {}
    operations: list[dict[str, Any]] = []
    for number, commit_value in full_cycles:
        if number == target_cycle and target_token:
            action, state = "set-commit", "ancestor"
            if commit_token(commit_value) == target_token:
                action, state = "none", "ancestor"
            else:
                changes[number] = target_token
            operations.append(
                {"cycle": number, "commit": target_token, "state": state, "action": action}
            )
            continue
        token = commit_token(commit_value)
        if token is None:
            operations.append(
                {"cycle": number, "commit": commit_value, "state": "exempt", "action": "none"}
            )
            continue
        state = ancestor_state(token, cwd)
        if state == "stale":
            changes[number] = "pending"
            operations.append(
                {"cycle": number, "commit": token, "state": "stale", "action": "reset-to-pending"}
            )
        else:
            operations.append({"cycle": number, "commit": token, "state": state, "action": "none"})

    sorted_changes = {n: v for n, v in sorted(changes.items(), reverse=True)}
    if not changes:
        return BackfillResult("clean", 0, operations, sorted_changes)
    if mode == "fix":
        return BackfillResult("fixed", 0, operations, sorted_changes)
    return BackfillResult("action-needed", 1, operations, sorted_changes)


def validate_progress_commits(content: str, cwd: Path | str = ".") -> list[str]:
    """Flag progress cycle ``commit`` hashes that are not ancestors of HEAD.

    A cycle ``commit`` must be ``pending``, ``N/A …``, or a commit that already
    exists in HEAD's history. A hash that resolves in the repository but is not
    an ancestor of HEAD is reported as a violation. Unknown hashes (different
    clone, shallow history) and non-git contexts are not flagged, to avoid false
    positives. YAML load failures and non-mapping roots return no commit
    violations (structural errors are reported separately by the validator).
    """
    try:
        data = load_yaml_mapping(content)
    except Exception:
        return []
    if not isinstance(data, dict):
        return []
    cycles = data.get("cycles")
    if not isinstance(cycles, list):
        return []
    if ancestor_state("HEAD", cwd) == "unavailable":
        return []
    violations: list[str] = []
    for entry in cycles:
        if not isinstance(entry, dict):
            continue
        raw = entry.get("commit")
        if not isinstance(raw, str):
            continue
        token = commit_token(raw)
        if token is None:
            continue
        if ancestor_state(token, cwd) == "stale":
            number = entry.get("number", "?")
            violations.append(
                f"progress: cycle {number} commit '{token}' is not an ancestor of HEAD "
                "(stale or self-referential); set it to `pending` or run "
                "`agentera check backfill --mode fix`, then forward-fill the product commit "
                "(never amend to backfill)"
            )
    return violations
